src/main/cpp/src/iceberg/iceberg_truncate.cu

```diff
@@ -0,0 +1,273 @@
+  char const* input_chars;
+  cudf::size_type const* input_offsets;
+  int32_t truncate_length;
+
+  // Output buffers
+  cudf::size_type* d_sizes;
+  char* d_chars;
+  cudf::detail::input_offsetalator d_offsets;
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    cudf::string_view str(input_chars + input_offsets[idx],
+                          input_offsets[idx + 1] - input_offsets[idx]);
+    if (!d_chars) {
+      // first phase
+      // Note: one character can be multiple(1-4) bytes for UTF8 encoding
+      if (str.length() < truncate_length) {
+        d_sizes[idx] = str.size_bytes();
+      } else {
+        auto tmp     = str.substr(0, truncate_length);
+        d_sizes[idx] = tmp.size_bytes();
+      }
+    } else {
+      // second phase
+      int out_len = d_offsets[idx + 1] - d_offsets[idx];
+      memcpy(d_chars + d_offsets[idx], str.data(), out_len);
+    }
+  }
+};
+
+struct truncate_binary_fn {
+  // Input data
+  char const* input_chars;
+  cudf::size_type const* input_offsets;
+  int32_t truncate_length;
+
+  // Output buffers
+  cudf::size_type* d_sizes;
+  char* d_chars;
+  cudf::detail::input_offsetalator d_offsets;
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    if (!d_chars) {
+      // first phase
+      int binary_len = input_offsets[idx + 1] - input_offsets[idx];
+      d_sizes[idx]   = (binary_len < truncate_length) ? binary_len : truncate_length;
+    } else {
+      // second phase
+      auto binary_ptr = input_chars + input_offsets[idx];
+      int out_len     = d_offsets[idx + 1] - d_offsets[idx];
+      memcpy(d_chars + d_offsets[idx], binary_ptr, out_len);
+    }
+  }
+};
+
+std::unique_ptr<cudf::column> truncate_integral_impl(cudf::column_view const& input,
+                                                     int32_t width,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(width != 0, "Width must not be zero");
+  if (input.is_empty()) { return cudf::make_empty_column(input.type().id()); }
+
+  auto output  = cudf::make_fixed_width_column(input.type(),
+                                              input.size(),
+                                              cudf::detail::copy_bitmask(input, stream, mr),
+                                              input.null_count(),
+                                              stream,
+                                              mr);
+  auto d_input = cudf::column_device_view::create(input, stream);
+
+  if (input.type().id() == cudf::type_id::INT32 || input.type().id() == cudf::type_id::DECIMAL32) {
+    // treat DECIMAL32 column as int32 column
+    using T = int32_t;
+    thrust::tabulate(rmm::exec_policy_nosync(stream),
+                     output->mutable_view().begin<T>(),
+                     output->mutable_view().end<T>(),
+                     truncate_integral_fn<T>{*d_input, width});
```

Comment:
I was wondering if there would be value in using a utility function to reduce the repetition of commonality:
```c++
template <typename RepT, typename Output>
void truncate_integral_and_fill(Output output, cuda::size_type num_rows, rmm::cuda_stream_view stream)
{
  auto const begin = output->mutable_view().begin<RepT>();
  thrust::fill_n(rmm::exec_policy_nosync(stream), begin, num_rows, truncate_integral_fn<RepT>{...};
}
```
Then the call sites will look a little shorter:
```c++
if (input.type().id() == cudf::type_id::INT32 || input.type().id() == cudf::type_id::DECIMAL32) {
  truncate_integral_and_fill<int32_t>(output, input.size(), stream);
}
```
Less boilerplate.

Reply:
Oh, right.  In my example, please replace `fill_n` with `tabulate`.  (`fill_n` doesn't take a generator function.)

My main point is to reduce that repetition, if possible.